Public documentation for CI Test monitors (#14073)
DataDog/documentation: content/en/monitors/create/types/ci.md

<div class="alert alert-warning">CI Visibility is not available in the selected site ({{< region-param key="dd_site_name" >}}) at this time.</div>
{{< /site-region >}}

<div class="alert alert-info">CI monitors are in alpha.
</div>

## Overview

Once [CI Visibility is enabled][1] for your organization, you can create a CI Pipeline monitor or a CI Test monitor.

CI monitors allow you to visualize CI data and set up alerts on it. For example, use a CI Pipeline monitor to receive alerts on a failed pipeline or a job, and use a CI Test monitor to receive alerts on failed tests or slow tests.

## Monitor creation

> To create a [CI Pipeline monitor][2] or [CI Test monitor][3] in Datadog, use the main navigation: *Monitors -> New Monitor --> CI*.

Reviewer: ```suggestion
To create a [CI monitor][2] in Datadog, use the main navigation: *Monitors -> New Monitor --> CI*.

Choose between a **Pipelines** or **Tests** monitor:
```

Similar to what APM has?

Author: indeed, it would be similar to APM monitor docs, but what link should we have to redirect the user to? `https://app.datadoghq.com/monitors/create/ci-pipelines`  or `https://app.datadoghq.com/monitors/create/ci-tests`?

Author: I guess it should be `https://app.datadoghq.com/monitors/create/ci-pipelines` since it's a first tab.

<div class="alert alert-info"><strong>Note</strong>: There is a default limit of 1000 CI monitors per account. <a href="/help/">Contact Support</a> to lift this limit for your account.</div>

> Choose between a **Pipelines** or **Tests** monitor:

Reviewer: ```suggestion
```

Author: why?

Reviewer: It's included in the previous comment.

<div class="alert alert-warning">CI Visibility is not available in the selected site ({{< region-param key="dd_site_name" >}}) at this time.</div>
{{< /site-region >}}

<div class="alert alert-info">CI monitors are in alpha.
</div>

## Overview

> Once [CI Visibility is enabled][1] for your organization, you can create a CI Pipeline or Test monitor.

Reviewer 2: For localization/translation
```suggestion
Once [CI Visibility is enabled][1] for your organization, you can create a CI Pipeline or CI Test monitor.
```

CI monitors allow you to visualize CI data and set up alerts on it. For example, create a CI Pipeline monitor to receive alerts on a failed pipeline or a job. Create a CI Test monitor to receive alerts on failed or slow tests.

## Monitor creation

To create a [CI monitor][2] in Datadog, use the main navigation: *Monitors -> New Monitor --> CI*.

<div class="alert alert-info"><strong>Note</strong>: There is a default limit of 1000 CI monitors per account. <a href="/help/">Contact Support</a> to lift this limit for your account.</div>

> Choose between a **Pipelines** or **Tests** monitor:

Reviewer 2: Agreed, this line needs to be removed, and the action statement should go at the beginning of the tab text in each section.

```suggestion
```

{{< tabs >}}
{{% tab "Pipelines" %}}

> 

Reviewer 2: ```suggestion

To create a CI Pipeline monitor:
```

{{< img src="monitors/monitor_types/ci_tests/define-the-search-query.png" alt="A query for CI Status:Error that is being set to group by Pipeline Name" style="width:80%;" >}}

#### Test runs with different parameters or configurations
If you run tests with the same test full name, but different test parameters or configurations, it's recommended to use `@test.fingerprint` in the monitor group by. This way alerts will trigger for test runs with specific test parameters or configurations. Using `@test.fignerpint` provides the same granularity level as the Test Stats, Failed and Flaky Tests section on the Commit Overview page.

For example, if a test with the same full name failed on Chrome, but passed on Firefox, then using the fingerprint will only trigger the alert on the Chrome test run.

> Using `@test.full_name` in this case might lead to incorrect behaviour.

Reviewer: ```suggestion
Using `@test.full_name` in this case will trigger the alert, even though the test passed on Firefox.
```

<div class="alert alert-info"><strong>Note</strong>: Only up to 2 queries can be used to build the evaluation formula per monitor.</div>

{{% /tab %}}
{{% tab "Tests" %}}

### Define the search query

1. Construct a search query using the same logic as a CI Test explorer search. For example, we can search failed tests for `main` branch using the following query: `@test.status:fail @git.branch:main`
2. Choose to monitor over a CI Test event count, facet, or measure:
    * **Monitor over a CI Test event count**: Use the search bar (optional) and do **not** select a facet or measure. Datadog evaluates the number of CI Test events over a selected time frame, then compares it to the threshold conditions.
    * **Monitor over a dimension**: If a dimension (qualitative facet) is selected, the monitor alerts over the `Unique value count` of the facet.

>     * **Monitor over a measure**: If a measure (quantitative facet) is selected, the monitor alerts over the numerical value of the CI Test facet (similar to a metric monitor) and aggregation needs to be selected (`min`, `avg`, `sum`, `median`, `pc75`, `pc90`, `pc95`, `pc98`, `pc99`, or `max`).

Author: I'm not sure as well 😄, it's without an article in https://docs.datadoghq.com/monitors/create/types/apm/?tab=analytics

>     * **Monitor over a measure**: If a measure (quantitative facet) is selected, the monitor alerts over the numerical value of the CI Test facet (similar to a metric monitor) and aggregation needs to be selected (`min`, `avg`, `sum`, `median`, `pc75`, `pc90`, `pc95`, `pc98`, `pc99`, or `max`).

Reviewer 2: Yep, should have an article in both places, I think the other instance was perhaps a mistake

3. Group CI Test events by multiple dimensions (optional):
   All CI Test events matching the query are aggregated into groups based on the value of up to four facets.
4. Configure the alerting grouping strategy (optional):

>    * If the query has a `group by`, you will receive an alert for every source according to the group parameters. An alerting event is generated for each group that meets the set conditions. For example, you could group a query by `@test.full_name` (which is a combination of a test bundle, test suite and test name, e.g. `MyBundle.MySuite.myTest`) to receive a separate alert for each CI Test full name when the number of errors is high.

Reviewer: I think `bundle` only applies to Swift tests. It could be confusing for other customers. Maybe rephrasing it?

{{< img src="monitors/monitor_types/ci_tests/define-the-search-query.png" alt="A query for CI Status:Error that is being set to group by Pipeline Name" style="width:80%;" >}}

#### Test runs with different parameters or configurations

> If you run tests with the same test full name, but different test parameters or configurations, it's recommended to use `@test.fingerprint` in the monitor group by. This way alerts will trigger for test runs with specific test parameters or configurations.

Reviewer: Maybe it'd be interesting to express here that this is the granularity level we have in the failed/flaky test tables.

For example, if a test with the same full name failed on Chrome browser, but it didn't fail on Firefox, the alert will be triggered for the test run on Chrome.
Using `@test.full_name` in this case might lead to incorrect behaviour.

> Test fingerpint is a hash value calculated using the following attributes:

Reviewer: Not sure if it's the best place to detail the fingerprint internals. This is not related to monitors exclusively. Also, it can change over time. If we really want to communicate the `test.fingerprint` feature, I'd suggest using a dedicated page in the CI section, and then linking it to that.

<div class="alert alert-info"><strong>Note</strong>: Only up to 2 queries can be used to build the evaluation formula per monitor.</div>

[1]: /continuous_integration/setup_pipelines/custom_commands/
{{% /tab %}}
{{% tab "Tests" %}}

> 

Reviewer 2: ```suggestion

To create a CI Test monitor:
```

### Define the search query

> 1. Construct a search query using the same logic as a CI Test explorer search. For example, we can search failed tests for a branch `main` using the following query: `@test.status:fail @git.branch:main`

Reviewer 2: ```suggestion
1. Construct a search query using the same logic as a CI Test explorer search. For example, you can search failed tests for a branch `main` using the following query: `@test.status:fail @git.branch:main`
```

2. Choose to monitor over a CI Test event count, facet, or measure:
    * **Monitor over a CI Test event count**: Use the search bar (optional) and do **not** select a facet or measure. Datadog evaluates the number of CI Test events over a selected time frame, then compares it to the threshold conditions.
    * **Monitor over a dimension**: If a dimension (qualitative facet) is selected, the monitor alerts over the `Unique value count` of the facet.
    * **Monitor over a measure**: If a measure (quantitative facet) is selected, the monitor alerts over the numerical value of the CI Test facet (similar to a metric monitor) and aggregation needs to be selected (`min`, `avg`, `sum`, `median`, `pc75`, `pc90`, `pc95`, `pc98`, `pc99`, or `max`).

3. Group CI Test events by multiple dimensions (optional):
   All CI Test events matching the query are aggregated into groups based on the value of up to four facets.
4. Configure the alerting grouping strategy (optional):

>    * If the query has a `group by`, you will receive an alert for every source according to the group parameters. An alerting event is generated for each group that meets the set conditions. For example, you could group a query by `@test.full_name` to receive a separate alert for each CI Test full name when the number of errors is high. Test full name is a combination of a test suite and test name, e.g. `MySuite.myTest`. In Swift test full name is a combination of a test bundle, suite and name, e.g. `MyBundle.MySuite.myTest`.

Reviewer 2: ```suggestion
   * If the query has a `group by`, an alert is sent for every source according to the group parameters. An alerting event is generated for each group that meets the set conditions. For example, you could group a query by `@test.full_name` to receive a separate alert for each CI Test full name when the number of errors is high. Test full name is a combination of a test suite and test name, for example: `MySuite.myTest`. In Swift, test full name is a combination of a test bundle, and suite and name, for example: `MyBundle.MySuite.myTest`.
```

{{< img src="monitors/monitor_types/ci_tests/define-the-search-query.png" alt="A query for CI Status:Error that is being set to group by Pipeline Name" style="width:80%;" >}}

#### Test runs with different parameters or configurations

> If you run tests with the same test full name, but different test parameters or configurations, it's recommended to use `@test.fingerprint` in the monitor group by. This way alerts will trigger for test runs with specific test parameters or configurations. Using `@test.fignerpint` provides the same granularity level as the Test Stats, Failed and Flaky Tests section on the Commit Overview page.

Reviewer 2: ```suggestion
If you run tests with the same test full name, but different test parameters or configurations, it's recommended to use `@test.fingerprint` in the monitor group by. This way, alerts trigger for test runs with specific test parameters or configurations. Using `@test.fignerpint` provides the same granularity level as the Test Stats, Failed and Flaky Tests section on the **Commit Overview** page.
```

> For example, if a test with the same full name failed on Chrome, but passed on Firefox, then using the fingerprint will only trigger the alert on the Chrome test run.

Reviewer 2: ```suggestion
For example, if a test with the same full name failed on Chrome, but passed on Firefox, then using the fingerprint only triggers the alert on the Chrome test run.
```

> Using `@test.full_name` in this case will trigger the alert, even though the test passed on Firefox.

Reviewer 2: ```suggestion

Using `@test.full_name` in this case triggers the alert, even though the test passed on Firefox.
```

#### Formulas and functions

You can create CI Test monitors using formulas and functions. For example, this can be used to create monitors on the **rate** of an event happening, such as the rate of a test failing (error rate).

The following example is a test error rate monitor using a formula that calculates the ratio of "number of failed test events" (`@test.status:fail`) over "number of total test events" (no filter), grouped by `@test.full_name` (to be alerted once per test).

> {{< img src="monitors/monitor_types/ci_tests/define-the-search-query-fnf.png" alt="Monitor being defined with steps a, b, and c, where steps a and b are queries and step c calculates the rate from them." style="width:80%;" >}}

Reviewer 2: Beautiful alt text, thank you! ❤️

> <div class="alert alert-info"><strong>Note</strong>: Only up to 2 queries can be used to build the evaluation formula per monitor.</div>

Reviewer 2: ```suggestion
<div class="alert alert-info"><strong>Note</strong>: A maximum of two queries can be used to build the evaluation formula per monitor.</div>
```

<div class="alert alert-info"><strong>Note</strong>: Only up to 2 queries can be used to build the evaluation formula per monitor.</div>

[1]: /continuous_integration/setup_pipelines/custom_commands/
{{% /tab %}}
{{% tab "Tests" %}}

To create a CI Test monitor:
### Define the search query

> 1. Construct a search query using the same logic as a CI Test explorer search. For example, you can search failed tests for a branch `main` using the following query: `@test.status:fail @git.branch:main`

Reviewer 3: Adding test service to this example

```suggestion
1. Construct a search query using the same logic as a CI Test explorer search. For example, you can search failed tests for the `main` branch of the `myapp` test service using the following query: `@test.status:fail @git.branch:main @test.service:myapp`
```
